lib/vdsm/virt/thinp.py

```diff
@@ -211,83 +215,108 @@ def on_block_threshold(self, target, path, threshold, excess):
                 'Unknown drive %r for vm %s - ignored block threshold event',
                 drive_name, self._vm.id)
         else:
-            drive.on_block_threshold(path)
+            if drive.on_block_threshold(path):
+                periodic.VolumeWatermarkMonitor.dispatch(self._vm)
+
+    def on_enospc(self, drive):
+        """
+        Called when a VM pauses because of ENOSPC error writing to drive.
+        """
+        if drive.on_enospc():
+            periodic.VolumeWatermarkMonitor.dispatch(self._vm)
 
     # Monitoring volumes.
 
-    def monitor_volumes(self):
+    def monitor_volumes(self, urgent=False):
         """
-        Return True if at least one volume is being extended, False otherwise.
+        Check and extend drives if needed.
+
+        If urgent is True, this is a call trigger by event from libvirt,
+        and we must extend some drives as soon as possible.  Always
+        False when called from normal periodic monitoring.
         """
         drives = self._monitored_volumes()
         if not drives:
-            return False
+            return
+
+        if not self._update_block_info(drives):
+            return
 
+        for drive in drives:
+            try:
+                self._extend_drive_if_needed(drive, urgent=urgent)
+            except ImprobableResizeRequestError:
+                break
```

Comment:
Why not continue to the next drive?

Reply:
> Why not continue to the next drive?

This is critical error, and the vm was already paused.